packages/controls/src/controls/group/group.ts

```diff
@@ -0,0 +1,310 @@
+  }
+
+  propsToData(propsData: PropsData<C>): DataType<C> {
+    return {
+      ...this.dataSignature,
+      value: propsData,
+    }
+  }
+
+  toData(value: ValueType<C>): DataType<C> {
+    return this.propsToData(
+      mapValues(this.propDefs, (def, key) => def.toData(value[key])),
+    )
+  }
+
+  copyData(
+    data: DataType<C> | undefined,
+    context: CopyContext,
+  ): DataType<C> | undefined {
+    if (data == null) return undefined
+
+    const propsData = Definition.propsData(data)
+
+    return this.propsToData(
+      mapValues(this.propDefs, (def, key) =>
+        def.copyData(propsData[key], context),
+      ),
+    )
+  }
+
+  getTranslatableData(data: DataType<C>): Data {
+    if (data == null) return null
+
+    const propsData = Definition.propsData(data)
+
+    return mapValues(this.propDefs, (def, key) =>
+      def.getTranslatableData(propsData[key]),
+    )
+  }
+
+  mergeTranslatedData(
+    data: DataType<C>,
+    translatedData: Record<string, DataType<C>>,
+    context: MergeTranslatableDataContext,
+  ): Data {
+    if (translatedData == null) return data
+
+    const propsData = Definition.propsData(data)
+
+    return mapValues(this.propDefs, (def, key) =>
+      def.mergeTranslatedData(propsData[key], translatedData[key], context),
+    )
+  }
+
+  createInstance(sendMessage: SendMessage): InstanceType<C> {
+    return new GroupControl(this, sendMessage)
+  }
+
+  serialize(): [SerializedRecord, Transferable[]] {
+    return serialize(this.config, {
+      type: Definition.type,
+    })
+  }
+
+  introspect<R>(data: DataType<C> | undefined, target: IntrospectionTarget<R>) {
+    if (data == null) return []
+
+    const propsData = Definition.propsData(data)
+
+    return Object.entries(this.propDefs).flatMap(
+      ([key, def]) => def.introspect(propsData[key], target) ?? [],
+    )
+  }
+
+  static propsData<C extends Config>(data: DataType<C>): PropsData<C> {
+    if (Definition.isVersionedData(data)) {
+      return data.value
+    }
+
+    return data
+  }
+
+  static propDataPath<C extends Config>(
+    data: DataType<C>,
+    key: string,
+  ): string[] {
+    return Definition.isVersionedData(data) ? ['value', key] : [key]
+  }
+
+  private static isVersionedData<C extends Config>(
+    data: DataType<C>,
+  ): data is VersionedData<C> {
+    return (
+      data != null &&
+      ControlDataTypeKey in data &&
+      (data[ControlDataTypeKey] === Definition.v1DataType ||
+        data[ControlDataTypeKey] === ShapeV2Definition.v1DataType)
+    )
+  }
+}
+
+export class GroupDefinition<C extends Config = Config> extends Definition<C> {}
+
+export function Group<D extends KeyDefinitions>(
+  config: Config<D>,
+): GroupDefinition<Config<D>> {
+  return new GroupDefinition<Config<D>>(
+    config?.preferredLayout == null
+      ? { preferredLayout: Definition.Layout.Inline, ...config }
+      : config,
+    Definition.v1DataType,
```

Comment:
`Group` constructor always constructs `GroupDefinition` with the group data type.